### ISSUE TITLE
Warn if unreachable code is provably reachable

### ISSUE THREAD

**Issue by dev-ardi** [rust-lang/rust#131167]
### Code

 if true { unsafe { unreachable_unchecked() } };
 if true { unreachable!() };

### Current output

No warnings

### Desired output

1. warning: this is ub if called...
2. warning: this will panic when called...

### Rust Version

rustc 1.83.0-nightly (c52c23b6f 2024-09-16)

**Comment by ryanavella** [comment]
Would you want it to warn on a custom wrapper?

```rust
fn my_unreachable() {
    unreachable!("a long message that you've factored out to a single function to reduce duplication")
}
```

**Comment by bjorn3** [comment]
And what if you call a function before calling `unreachable!()`? It may be the case that this function is known to never return given the arguments that are used at this place.

**Comment by dev-ardi** [comment]
> Would you want it to warn on a custom wrapper?

I don't think that we want that, from a simplicity of analysis POV. If the user wants to avoid repetition they should probably just factor out the string.

> what if you call a function before calling `unreachable!()`? 

I don't know whether we have infrastructure/queries in place to check whether a function might diverge. Admittedly this is a hard problem and it is computationally expensive, but it might aid in many other lints.

With how many functions have a branch that panics somewhere deep on the stack I think that *most* functions might diverge actually :/

**Comment by Ved-s** [comment]
For now there's a cursed way of achieving this,
```rust
#[deny(unfulfilled_lint_expectations)]
pub fn unreachable_test() {
    return;

    #[expect(unreachable_code)]
    {}
}
```
```rust
#[deny(unfulfilled_lint_expectations)]
pub fn unreachable_test() {
    // return;

    #[expect(unreachable_code)] //error: this lint expectation is unfulfilled
    {}
}
```

But a warning on reachable `unreachable!()` or a separate macro that errors when it's reachable would still be useful.

upd: you can use two lint attributes on the same item
```rust
#[deny(unfulfilled_lint_expectations)]
#[expect(unreachable_code)]
{}
```
